### PR TITLE
Ruby on Rails - Recipe App (compare main from dev branch). (Team of two members).

### PR DESCRIPTION
**Pull request by jcmunav63**
### Description
The Recipe app keeps track of all your recipes, ingredients, and inventory. It will allow you to save ingredients, keep track of what you have, create recipes, and generate a shopping list based on what you have and what you are missing from a recipe. Also, since sharing recipes is an important part of cooking the app should allow you to make them public so anyone can access them.

### How to build the Recipe app
The app will have some common interfaces, but depending on your team size it will have a couple of extra ones.

- A login page.
- A registration page.
- A recipes list (with all CRUD implementation, except for 'update').
- A list of all public recipes from other users with their names and total prices.

If your team has fewer than 3 members, you should implement:
- A food list (with all CRUD implementation, except for 'update').
- A general shopping list view (all missing food for all your recipes and total price).

The wireframes below are only for teams with less than 3 people (check this document: [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/recipe-app/buisness_requirements.md))
- For the below wireframe image (Food list), make sure to implement also a column "Quantity" to display how much food you have for each ingredient.

### General requirements
- Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
- Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

### Ruby requirements
- Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

### Project requirements
- You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.

- Login page and registration page:

   - Should be built with Devise.
- Recipes list:
  - Should display a list of recipes created by the logged-in user as in the wireframe.
  - Should lead to recipe details.
- Public recipe list:

  - Should display a list of all public recipes ordered by newest as in the wireframe.
  - Should lead to recipe details.
  - If the user is the owner of the recipe, should allow the user to delete it.
- Recipe details:
  - Should display a toggle button that allows for a recipe to be made public or private.
  - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
  - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
- Make sure there are no N+1 queries happening.
- Create a navigation menu that allows users to open all of the pages you created.

- Write unit and integration tests

- Only for a group that has less than 3 members:
- Food list:
  - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
  - Should lead to a form that allows users to add new food.
- General shopping list view:
  - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
  - Should count the total food items and total price of the missing food.

### Technical setup
- Set up the repository on GitHub and use Gitflow.
- Set up Devise for authentication.
- Set up RSpec and Capybara testing libraries.

### CODE REVIEW

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    @recipefoods = @recipe.recipe_foods
+  end
+
+  def generate_shopping_list
+    @recipe = Recipe.find(params[:id])
+    @missing_food_items = find_missing_food_items(@recipe)
+    @total_food_to_buy = @missing_food_items.count
+    @total_value_of_food_needed = @missing_food_items.sum { |item| item[:quantity] * item[:price] }
+    @missing_food_items = Food.where(id: @missing_food_items.map(&:id)).includes(:user) # Preload users
+  end
```
Comment (OyePriscilla) [inline app/controllers/recipes_controller.rb:52]: - [ ] Please kindly ensure that generate_shopping_list is also part of the navbar items as state in the requirement. 🙏 

Untitled

File: config/routes.rb [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+  get 'recipes/index'
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
   # Reveal health status on /up that returns 200 if the app boots with no exceptions, otherwise 500.
   # Can be used by load balancers and uptime monitors to verify that the app is live.
-  get "up" => "rails/health#show", as: :rails_health_check
+  get 'up' => 'rails/health#show', as: :rails_health_check
 
   # Defines the root path route ("/")
   # root "posts#index"
```
Comment (OyePriscilla) [inline config/routes.rb:12]: - [ ] Please kindly remove this commented code in order to make your work neater. ✔️

File: app/views/shared/_menu.html.erb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    <% if user_signed_in? %>
+      <%= button_to 'Logout', destroy_user_session_path, method: :delete, class: "logout-btn" %>
+    <% end %>
+  </span>
+  <ul class="menu">
+    <li><%= link_to "Recipes", root_path, class: "menu-btn" %></li>
+    <li><%= link_to "Public Recipes", public_recipes_path, class: "menu-btn" %></li>
+    <li><%= link_to "Foods", user_foods_path(current_user), class: "menu-btn" %></li>
+    <li><a class="menu-btn" href="#">About</a></li>
+  </ul>
```
Comment (OyePriscilla) [inline app/views/shared/_menu.html.erb:13]: - [optional] I think you can make this `About` menu item to be functional in order to make it to be interactive. 👍

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @jcmunav63 and @mohamedgamaleldin999999 👏 👏 👏 👏, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

so-close

## Highlights
Use ruby gems as software packages system.
Install Ruby on Rails framework.
Use controllers to handle requests and render empty views.
Use params from browser request in a safe way.
Use preprocessed html file with embedded Ruby code.
Use layouts and templates for shared content.
Use database migration files to maintain database schema.
Use validations for models.
Set up associations between models.
Build a webapp that requires the user to log in.
Use devise gem for authentication.

## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by 1GORDON — APPROVED** [review]
# Status: APPROVED 🟢 

Hi @jcmunav63, and @mohamedgamaleldin999999 ✋

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! :tada: :clap: :clap: :100:

## Highlights!
- Implemented requested changes ✔️

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**